File: src/aizk/graph/reembed.py
from itertools import batched

from loguru import logger
from sqlalchemy import update
from sqlmodel import select

from ..config import settings
from ..serving import embed
from ..store import Chunk, Community, EntityContent, FactContent, Profile
from ..store.engine import Session, bypass_rls
from ..types import Scopes

ScopedEmbedded = Chunk | Community | Profile

_SCOPED_TARGETS: dict[type[ScopedEmbedded], str] = {
    Chunk: "text",
    Community: "summary",
    Profile: "summary",
}

ContentEmbedded = EntityContent | FactContent
_CONTENT_TARGETS: dict[type[ContentEmbedded], str] = {
    EntityContent: "name",
    FactContent: "statement",
}

EmbeddedTable = ScopedEmbedded | ContentEmbedded


async def rewrite_embeddings(
    session: Session,
    model: type[EmbeddedTable],
    field: str,
    scopes: Scopes | None = None,
) -> int:
    """Re-read one table's source text and overwrite its embedding column in batches."""
    column = getattr(model, field)
    selection = select(model.id, column).order_by(model.id)
    if scopes is not None:
        selection = selection.where(model.__table__.c.scopes == sorted(scopes))
    rows = (await session.exec(selection)).all()
    for batch in batched(rows, settings.reembed_batch, strict=False):
        vectors = await embed([source for _, source in batch], mode="document")
        await session.exec(
            update(model),
            params=[
                {"id": row_id, "embedding": vector}
                for (row_id, _), vector in zip(batch, vectors, strict=True)
            ],
        )
    return len(rows)


async def reembed_scoped_table(
    scopes: Scopes,
    model: type[ScopedEmbedded],
    field: str,
) -> int:
    """Re-embed one per-tenant table's rows under the acting user, return the count
    rewritten."""
    async with bypass_rls() as session:
        count = await rewrite_embeddings(session, model, field, scopes)
    logger.info("re-embedded {} {} rows", count, model.__tablename__)
    return count


async def reembed_content_table(
    model: type[ContentEmbedded],
    field: str,
) -> int:
    """Re-embed every row of one deduplicated content table, return how many vectors changed."""
    async with bypass_rls() as session:
        count = await rewrite_embeddings(session, model, field)
    logger.info("re-embedded {} {} content rows", count, model.__tablename__)
    return count


async def reembed(scopes: Scopes | None = None) -> int:
    """Re-encode every stored embedding with the current embedder, the one-command model
    migration."""
    key = frozenset(scopes or (settings.system_user_id,))
    scoped = sum(
        [await reembed_scoped_table(key, model, field) for model, field in _SCOPED_TARGETS.items()]
    )
    content = sum(
        [await reembed_content_table(model, field) for model, field in _CONTENT_TARGETS.items()]
    )
    return scoped + content

File: archive/ts/src/lib/server/graph/reembed.ts
// The one-command embedding-model migration, the TS twin of graph/reembed.py: re-read each
// embedded table's source text and overwrite its embedding column in batches, all under
// the owner role since content tables carry no update policy and community rows no update
// policy either. Scoped tables rewrite one exact scope set; content tables rewrite fully.
import { sql, type SQL } from 'drizzle-orm';
import { chunk as batched } from 'es-toolkit';

import { embed } from '../serving';
import { settings } from '../settings';
import { bypassRls, scopeKey, uuidArray, type Tx } from './system';

export interface EmbeddedTarget {
	table: SQL;
	source: SQL;
	scoped: boolean;
}

export const EMBEDDED_TARGETS: EmbeddedTarget[] = [
	{ table: sql.raw('chunk'), source: sql.raw('text'), scoped: true },
	{ table: sql.raw('community'), source: sql.raw('summary'), scoped: true },
	{ table: sql.raw('profile'), source: sql.raw('summary'), scoped: true },
	{ table: sql.raw('entity_content'), source: sql.raw('name'), scoped: false },
	{ table: sql.raw('fact_content'), source: sql.raw('statement'), scoped: false }
];

// Re-read one table's source text and overwrite its embedding column in batches, the
// rewrite_embeddings twin, returning how many rows were rewritten.
export const rewriteEmbeddings = async (
	tx: Tx,
	target: EmbeddedTarget,
	scopes?: string[]
): Promise<number> => {
	const filter = scopes ? sql` where scopes = ${uuidArray(scopes)}::uuid[]` : sql``;
	const rows = (await tx.execute(sql`
		select id, ${target.source} as source from ${target.table}${filter} order by id
	`)) as unknown as { id: string; source: string }[];
	for (const batch of batched(rows, settings.reembedBatch)) {
		const vectors = await embed(batch.map((row) => row.source), 'document');
		const payload = JSON.stringify(
			batch.map((row, index) => ({ id: row.id, embedding: vectors[index] }))
		);
		await tx.execute(sql`
			update ${target.table} t set embedding = (d.embedding)::halfvec
			from jsonb_to_recordset(${payload}::jsonb) as d(id uuid, embedding text)
			where t.id = d.id
		`);
	}
	return rows.length;
};

// Re-encode every stored embedding with the current embedder, the reembed twin: one
// owner-role transaction per table, the per-tenant tables filtered to the scope key.
export const reembed = async (scopes?: string[]): Promise<number> => {
	const key = scopeKey(scopes);
	let total = 0;
	for (const target of EMBEDDED_TARGETS)
		total += await bypassRls((tx) =>
			rewriteEmbeddings(tx, target, target.scoped ? key : undefined)
		);
	return total;
};
